features/steps/thens/templates.py

```diff
@@ -0,0 +1,82 @@
+from collections import defaultdict
+import functools
+import pprint
+from behave import *
+
+import errors as err
+from features.steps.utils import misc
+
+def group_template_tuples(context):
+    if instances := getattr(context, 'instances', None):
+        values = [[d for d in context.template_tuples if inst in d.values()] for inst in instances]
+    else:
+        values = [context.template_tuples]
+    return values
+
+def group_template_tuples_by_root(context):
+    grouping = defaultdict(list)
+    for d in context.template_tuples:
+        grouping[d.get('_root')].append(d)
+    return [(vs[0].get('_root'), vs) for vs in grouping.values()]
+
+
+def grouped_templates_tuples(func):
+    @functools.wraps(func)
+    def inner(context, *args, **kwargs):
+        if hasattr(context, 'instances'):
+            grouping = zip(context.instances, group_template_tuples(context))
+        else:
+            grouping = group_template_tuples_by_root(context)
+
+        func(context, grouping, *args, **kwargs)
+    return inner
+
+@then('all "{key}" equals "{value}"')
+@then('all "{key}" equals "{value}" or "{value_2}"')
+@grouped_templates_tuples
+def step_impl(context, grouping, key, value, value_2 = None):
+    required = {value}
+    if value_2 is not None:
+        required.add(value_2)
+
+    errors = []
+
+    if context.template_tuples:
+        for inst, vs in grouping:
+            if vs:
+                values = set(d.get(key) for d in vs)
+                invalid = values - required
+                if invalid:
+                    errors.append(err.TemplateValuationError(inst, f"{' '.join(invalid)} do not match {' or '.join(required)}"))
```

Comment:
```suggestion
                values = set(d.get(key) for d in vs)
                invalid = values - required
                if invalid:
                    errors.append(err.TemplateValuationError(inst, f"{' '.join({str(item) for item in invalid})} does not match {' or '.join(required)}"))
```

The absence of labels in `fail-abc999-UT_AWC_1_no_geometry_ifc4x3tc1_no_labels ` resulted in
`invalid = values - required = {None} - {'StartsAt', 'EndsAt'} = {None}`

which cannot be appended to the errors 
```
' '.join(invalid) # invalid = {None}
                values = set(d.get(key) for d in vs)
                invalid = values - required
                if invalid:
                    errors.append(err.TemplateValuationError(inst, f"{' '.join(invalid)} do not match {' or '.join(required)}"))
Traceback (most recent call last):
TypeError: sequence item 0: expected str instance, NoneType found
```

With the suggestion, the output is : 
```
Assertion Failed: Errors occured:
      On instance #33=IfcAlignmentHorizonta...,$,$,$) there is a template error. None do not match StartsAt or EndsAt
```